xbmc/guilib/test/TestGUIControlFactory.cpp

```diff
@@ -0,0 +1,665 @@
+                     {UTILS::MOVING_SPEED::EventType::NONE,
+                      UTILS::MOVING_SPEED::EventCfg{1.0, 2.0, 3, 5.0}},
+                    }},
+    MovingSpeedTest{R"(<root><test>
+                         <eventconfig type="up"/>
+                         <eventconfig type="down" acceleration="2.0"/>
+                         <eventconfig type="left" maxvelocity="3.0"/>
+                         <eventconfig type="right" resettimeout="4"/>
+                         <eventconfig type="none" delta="5"/>
+                       </test></root>)"s,
+                    {{UTILS::MOVING_SPEED::EventType::UP,
+                      UTILS::MOVING_SPEED::EventCfg{0.0, 0.0, 0, 0.0}},
+                     {UTILS::MOVING_SPEED::EventType::DOWN,
+                      UTILS::MOVING_SPEED::EventCfg{2.0, 0.0, 0, 0.0}},
+                     {UTILS::MOVING_SPEED::EventType::LEFT,
+                      UTILS::MOVING_SPEED::EventCfg{0.0, 3.0, 0, 0.0}},
+                     {UTILS::MOVING_SPEED::EventType::RIGHT,
+                      UTILS::MOVING_SPEED::EventCfg{0.0, 0.0, 4, 0.0}},
+                     {UTILS::MOVING_SPEED::EventType::NONE,
+                      UTILS::MOVING_SPEED::EventCfg{0.0, 0.0, 0, 5.0}},
+                    }},
+    MovingSpeedTest{R"(<root><test/></root>)"s,
+                    {}, true},
+    MovingSpeedTest{R"(<root/>)"s,
+                    {}, false},
+};
+
+struct PositionTest
+{
+  std::string def;
+  float pos;
+  bool result = true;
+};
+
+class TestGetPosition : public testing::WithParamInterface<PositionTest>, public testing::Test
+{
+};
+
+const auto PositionTests = std::array{
+    PositionTest{R"(<root><test>0.1</test></root>)"s, 0.1f},
+    PositionTest{R"(<root><test>0.2r</test></root>)"s, -0.1f},
+    PositionTest{R"(<root><test>0.2%</test></root>)"s,
+                 0.2f * 0.1f / 100.f},
+    PositionTest{R"(<root><test>something</test></root>)"s, 0.f},
+    PositionTest{R"(<root><test/></root>)"s, 0.f, false},
+    PositionTest{R"(<root/>)"s, 0.f, false},
+};
+
+struct StringTest
+{
+  std::string def;
+  std::string value;
+  bool result = true;
+};
+
+class TestGetString : public testing::WithParamInterface<StringTest>, public testing::Test
+{
+};
+
+const auto StringTests = std::array{
+    StringTest{R"(<root><test>foo</test></root>)"s, "foo"s},
+    StringTest{R"(<root><test>1</test></root>)"s, "Pictures"s},
+    StringTest{R"(<root><test>-1</test></root>)"s, "-1"s},
+    StringTest{R"(<root><test/></root>)"s, ""s, true},
+    StringTest{R"(<root/>)"s, ""s, false},
+};
+
+struct TypeTest
+{
+  std::string def;
+  std::string type;
+};
+
+class TestGetType : public testing::WithParamInterface<TypeTest>, public testing::Test
+{
+};
+
+const auto TypeTests = std::array{
+    TypeTest{R"(<root type="foo"/></root>)"s, "foo"s},
+    TypeTest{R"(<root><type>foo</type></root>)"s, "foo"s},
+    TypeTest{R"(<root/>)"s, ""s},
+};
+
+}
```

Comment:
This end brace confused me until I remembered the anonymous namespace. I thought clang-format automatically added `} // namespace` but I guess not so it couldn't hurt to add this manually.

Reply:
clang-format is going to piss its yellow stuff all across my commits anyways so it will come in then.